File: 0498-diagonal-traverse/0498-diagonal-traverse.java
class Solution {
    public int[] findDiagonalOrder(int[][] mat) {

        Map<Integer, List<Integer>> map = new HashMap<>();
        int m = mat.length;
        int n = mat[0].length;
        int[] ans = new int[m * n];

        // Step 1: Store diagonals in the map
        for (int i = 0; i < mat.length; i++) {
            for (int j = 0; j < mat[0].length; j++) {
                if (!map.containsKey(i + j)) {
                    map.put(i + j, new ArrayList<>());
                }
                map.get(i + j).add(mat[i][j]);
            }
        }

        // Step 2: Traverse the map and reverse alternative diagonals
        int index = 0;
        // boolean reverse = false; // \U0001f448 Moved outside the loop

        for (Map.Entry<Integer, List<Integer>> entry : map.entrySet()) {
            List<Integer> list = entry.getValue();

            if (entry.getKey() % 2 == 0) {
                Collections.reverse(list);
            }
            for (int i = 0; i < list.size(); i++) {
                ans[index++] = list.get(i);
            }

            // reverse = !reverse; // \U0001f448 Toggle the flag
        }

        return ans;

    }
}

File: 0498-diagonal-traverse/0498-diagonal-traverse.py
class Solution:
    def findDiagonalOrder(self, mat: List[List[int]]) -> List[int]:
        map={}
        result=[]
        for i in range(0,len(mat)):
            for j in range(0,len(mat[0])):
                if i+j not in map:
                    map[i+j]=[]
                map[i+j].append(mat[i][j])
                
        for key,values in map.items():
            if key !=0 and key%2==0:
                values.reverse()
                result.extend(values)
            else:
                result.extend(values)
        print(result)

        return result


                

        